compiler/src/dotty/tools/dotc/util/ReusableInstance.scala

```diff
@@ -0,0 +1,37 @@
+package dotty.tools.dotc.util
+
+import scala.collection.mutable.ArrayBuffer
+import scala.util.chaining._
+
+/** A wrapper for a list of cached instances of a type `T`.
+  * The wrapper is recursion-reentrant: several instances are kept, so
+  * at each depth of reentrance we are reusing the instance for that.
+  *
+  * An instance is created upon creating this object, and more instances
+  * are allocated dynamically, on demand, when reentrance occurs.
+  *
+  * Not thread safe.
+  *
+  * Ported from scala.reflect.internal.util.ReusableInstance
+  */
+final class ReusableInstance[T <: AnyRef] private (make: => T, enabled: Boolean) {
+  private[this] val cache = if (enabled) new ArrayBuffer[T](ReusableInstance.InitialSize).tap(_.addOne(make)) else null
+  private[this] var taken = 0
+
+  @inline def using[R](action: T => R): R =
```

Comment:
Since `using` is a keyword in dotty I think it's worth diverging from scala 2 and using a different name here (also can we make it inline, and can we drop the `enabled` parameter since it's always true?)